server/static/dashboard/live.js

```diff
@@ -0,0 +1,141 @@
+    var item = document.importNode(document.querySelector('#collab').content, true);
+    var root = item.querySelector('.collab');
+    root.setAttribute('id', 'collab-' + collab.id);
+    let users = collab.data.users.slice().sort();
+    root.dataset.users = users.join(',');
+    var link = item.querySelector('a');
+    var href = '/dashboard/' + project + '/' + collab.id;
+    link.setAttribute('href', href);
+    link.textContent = users.join('\n');
+    list.insertBefore(item, list.children[atIndex + idx]);
+
+    // TODO: move into view file
+    var codeline = document.createElement('span');
+    codeline.style.fontSize = '8px';
+    codeline.style.background = 'rgba(252, 248, 227)';
+    root.querySelector('.panel-body').appendChild(document.createElement('br'));
+    root.querySelector('.panel-body').appendChild(codeline);
+
+    connection.createFetchQuery('files', { collabid: collab.id }, {}, function(err, files) {
+      if (err) { throw err; }
+
+      var typingtimeoutid = undefined;
+      function updateTypingDisplay() {
+        root.classList.add('typing');
+        clearTimeout(typingtimeoutid);
+        typingtimeoutid = setTimeout(function() {
+          root.classList.remove('typing');
+        }, 1000);
+      }
+
+      function updateErrorDisplay() {
+        var maxlevel = 0;
+        files.forEach(function(file) {
+          for (var k in file.data.markers) {
+            file.data.markers[k].forEach(function(marker) {
+              maxlevel = Math.max(maxlevel, marker.severity);
+            });
+          }
+        });
+        root.classList.remove('error');
+        root.classList.remove('warning');
+        if (maxlevel == 2) root.classList.add('error');
+        else if (maxlevel == 1) root.classList.add('warning');
+      }
+
+      function updateMatchDisplay() {
+        if (searchregex) {
+          var match = false;
+          codeline.innerText = '';
+          files.forEach(function(file) {
+            var matchdata = file.data.text.match(searchregex);
+            if (matchdata) {
+              match = true;
+              codeline.innerText = findLineForIndex(matchdata.input, matchdata.index);
+            }
+          });
+          if (match) root.classList.add('match');
+          else root.classList.remove('match');
+          countdisplay.innerHTML = document.getElementsByClassName('match').length
+                           + '/' + document.getElementsByClassName('connected').length;
+        } else {
+          root.classList.remove('match');
+          codeline.innerText = '';
+        }
+      }
+
+      searchbox.addEventListener('input', updateMatchDisplay);
+
+      files.forEach(function(file) {
+        file.subscribe(function() {
+          root.classList.remove('disconnected');
+          root.classList.add('connected');
+          updateMatchDisplay();
+          updateErrorDisplay();
+          file.on('op', function(op) {
+            var optype = op[0].p[0];
```

Comment:
Rename

Reply:
Always update everything?

Reply:
This would mean that marker updates (from saving files) will trigger typing indicators, which doesn't seem desired

Reply:
Discussion: this is fine, since marker updates only happen in close proximity to actual changes (and can themselves be considered "actions", similar to typing)